File: 11724.py
import sys

sys.setrecursionlimit(10000)


node, edge = map(int, sys.stdin.readline().split())

graph = [[] for _ in range(node+1)]

visited = [False for _ in range(node+1)]

for _ in range(edge): #making graph
    n1, n2 = map(int, sys.stdin.readline().split())
    graph[n1].append(n2)
    graph[n2].append(n1)
    graph[n1].sort()
    graph[n2].sort()

    
def DFS(graph, visited, node):
    visited[node] = True
    for i in graph[node]:
        if not visited[i]:
            DFS(graph,visited,i)



comp = 0


for i in range(1,node+1):
    if visited[i] == False:
        DFS(graph,visited,i)
        comp += 1
        
sys.stdout.write(str(comp))

File: 11724.cpp
#include <iostream>
#include <vector>
using namespace std;

void DFS(vector<int> *graph, bool *visited, int node){
	visited[node]=true;
	for(int i=0;i<graph[node].size();i++){
		int next = graph[node][i];
		cout<<"node"<<node<<"next"<<next;
		if(visited[next]==false){		
			DFS(graph,visited,next);
		}
	}
}

int main(){
	int node, edge;
	cin>>node>>edge;
	vector<int> graph[node+1];
	
	for(int i=0;i<edge;i++){
		int n1, n2;
		cin>>n1>>n2;
		graph[n1].push_back(n2);
		graph[n2].push_back(n1);
	}
	
	bool visited[node+1]={false};
	int comp = 0;
	for(int i=1;i<=node;i++){
		if(visited[i]==false){
			DFS(graph,visited,i);
			comp += 1;	
		}
	}
	
	cout<<comp;
}
